### PR TITLE
Introduce RetryBackoffType.ExponentialWithJitter

### LINKED ISSUE #1108: Introduce RetryBackoffType.ExponentialWithJitter retry delay
In the #1093 we introduce a new `RetryBackoffType` enum member. We should add a `RetryBackoffType.ExponentialWithJitter` enum member to it, update `RetryHelper` class with new delay generation and make the `ExponentialWithJitter` the default one. 

We should consider using the code from [Polly.Contrib.WaitAndRetry](https://github.com/Polly-Contrib/Polly.Contrib.WaitAndRetry/blob/master/src/Polly.Contrib.WaitAndRetry/Backoff.DecorrelatedJitterV2.cs) project for this. 

See #530 for more details about the benefits of jitter delay.

### PR DESCRIPTION
**Pull request by martintmk**
### The issue or feature being addressed

Closes #1108 

### Details on the issue fix or feature implementation

Adding `RetryBackoffType.ExponentialWithJitter` retry delay type and making it a default one.

Adopted from:
https://github.com/Polly-Contrib/Polly.Contrib.WaitAndRetry/blob/master/src/Polly.Contrib.WaitAndRetry/Backoff.DecorrelatedJitterV2.cs

### Confirm the following

- [x]  I started this PR by branching from the head of the default branch
- [x]  I have targeted the PR to merge into the default branch
- [x]  I have included unit tests for the issue/feature
- [x]  I have successfully run a local build

### CODE REVIEW

File: src/Polly.Core/Retry/RetryHelper.cs [hunk 1]
```diff
@@ -25,7 +25,53 @@ public static TimeSpan GetRetryDelay(RetryBackoffType type, int attempt, TimeSpa
...
 #endif
+            RetryBackoffType.ExponentialWithJitter => DecorrelatedJitterBackoffV2(attempt, baseDelay, ref state, random),
             _ => throw new ArgumentOutOfRangeException(nameof(type), type, "The retry backoff type is not supported.")
         };
     }
+
+    /// <summary>
+    /// Generates sleep durations in an exponentially backing-off, jittered manner, making sure to mitigate any correlations.
+    /// For example: 850ms, 1455ms, 3060ms.
+    /// Per discussion in Polly issue 530, the jitter of this implementation exhibits fewer spikes and a smoother distribution than the AWS jitter formula.
```
Comment (martincostello) [inline src/Polly.Core/Retry/RetryHelper.cs:36]: Lets make "Polly issue 530" a link to the actual issue?

File: src/Polly.Core/Retry/RetryHelper.cs [hunk 2]
```diff
@@ -25,7 +25,53 @@ public static TimeSpan GetRetryDelay(RetryBackoffType type, int attempt, TimeSpa
...
+    /// </summary>
+    /// <param name="attempt">The current attempt.</param>
+    /// <param name="baseDelay">The median delay to target before the first retry, call it f (= f * 2^0).
+    /// Choose this value both to approximate the first delay, and to scale the remainder of the series.
+    /// Subsequent retries will (over a large sample size) have a median approximating retries at time f * 2^1, f * 2^2 ... f * 2^t etc for try t.
+    /// The actual amount of delay-before-retry for try t may be distributed between 0 and f * (2^(t+1) - 2^(t-1)) for t >= 2;
+    /// or between 0 and f * 2^(t+1), for t is 0 or 1.</param>
```
[the remark this answers is not in the document]
Comment (martintmk) [reply src/Polly.Core/Retry/RetryHelper.cs]: Done.

File: src/Polly.Core.Tests/Utils/RandomUtilTests.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+        var util = new RandomUtil(seed);
+
+        util.Invoking(u => u.NextDouble()).Should().NotThrow();
+    }
+
+    [Fact]
+    public void Instance_Ok()
+    {
+        RandomUtil.Instance.Should().NotBeNull();
+    }
```
[the remark this answers is not in the document]
Comment (martintmk) [reply src/Polly.Core.Tests/Utils/RandomUtilTests.cs:22]: Good idea, added the test.

File: src/Polly.Core/Retry/RetryHelper.cs (hunk 1 above)
Comment (martintmk) [reply src/Polly.Core/Retry/RetryHelper.cs]: Done.

File: src/Polly.Core/Utils/RandomUtil.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+#pragma warning disable CA5394 // Do not use insecure randomness
+#pragma warning disable S2931 // Classes with "IDisposable" members should implement "IDisposable"
+
+internal sealed class RandomUtil
+{
+    private readonly ThreadLocal<Random> _random;
+
+    public static readonly RandomUtil Instance = new(null);
+
+    public RandomUtil(int? seed) => _random = new ThreadLocal<Random>(() => seed == null ? new Random() : new Random(seed.Value));
```
[the remark this answers is not in the document]
Comment (martintmk) [reply src/Polly.Core/Utils/RandomUtil.cs:15]: I tried this, but the code still needs ifdefs making it less readable. I would prefer to leave it as it is.

**Review by martincostello — APPROVED** [review]

**Pull request merged**